### PR TITLE
First stab at doing codegen for the FSA

### PR DESCRIPTION
**Pull request by julienledem**
requires to checkout brennus [1] and "mvn install"
[1] https://github.com/julienledem/brennus

### CODE REVIEW

File: redelm-column/src/main/java/redelm/column/mem/MemColumnReader.java [hunk 1]
```diff
@@ -106,8 +107,12 @@ public int getCurrentDefinitionLevel() {
   }
 
   private void read() {
-    repetitionLevel = repetitionLevelColumn.readInteger();
-    definitionLevel = definitionLevelColumn.readInteger();
+    if (!(repetitionLevelColumn instanceof DevNullColumnReader)) {
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/column/mem/MemColumnReader.java]: Is this another performance thing? Do you really think the cost of a cached method call is greater than doing an instanceof?

File: redelm-column/src/main/java/redelm/io/RecordReader.java [hunk 2]
```diff
@@ -1,304 +1,18 @@
-/**
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/io/RecordReader.java:1]: No more license?

File: redelm-column/src/main/java/redelm/io/RecordReaderCompiler.java [hunk 3]
```diff
@@ -0,0 +1,386 @@
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/io/RecordReaderCompiler.java:0]: license etc

File: redelm-column/src/main/java/redelm/io/RecordReaderCompiler.java [hunk 4]
```diff
@@ -0,0 +1,386 @@
...
+import brennus.StatementBuilder;
+import brennus.SwitchBuilder;
+import brennus.ThenBuilder;
+import brennus.asm.ASMTypeGenerator;
+import brennus.model.FutureType;
+import brennus.printer.TypePrinter;
+
+public class RecordReaderCompiler {
+
+  public static abstract class BaseRecordReader<T> extends RecordReader<T> {
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/io/RecordReaderCompiler.java]: Is there any reason this shouldn't be in another file? A 100+ line class seems worth of its own file

File: redelm-column/src/main/java/redelm/column/mem/MemColumnReader.java (hunk 1 above)
Comment (julienledem) [inline redelm-column/src/main/java/redelm/column/mem/MemColumnReader.java]: It is more to see the impact in the result of profiling. DevNullColumnReader.readInteger() shows disproportionate high times while profiling but I suspect it is an artifact of JIT compilation messing up profiling. I'm currently thinking of decoupling those three, remove the read() method in MemColumnReader and just call the individual 3 columns as needed in the RecordReader. The compiled FSA would not call the columns at all when the max level is 0.

File: redelm-column/src/main/java/redelm/io/RecordReader.java (hunk 2 above)
Comment (julienledem) [inline redelm-column/src/main/java/redelm/io/RecordReader.java:1]: ah yes. will do

File: redelm-column/src/main/java/redelm/io/RecordReaderCompiler.java (hunk 4 above)
Comment (julienledem) [inline redelm-column/src/main/java/redelm/io/RecordReaderCompiler.java]: agreed. This is still a little WIP. I created to Pull Request to start sharing as this is the first version that works and improves a little bit perf.

File: redelm-column/src/main/java/redelm/column/mem/MemColumnReader.java [hunk 5]
```diff
@@ -106,8 +107,12 @@ public int getCurrentDefinitionLevel() {
...
+      repetitionLevel = repetitionLevelColumn.readInteger();
+    }
+    if (!(definitionLevelColumn instanceof DevNullColumnReader)) {
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/column/mem/MemColumnReader.java]: I know that you are planning on changing some of this stuff around. Can you a) include some TODO comments and b) update the issues page?

File: redelm-column/src/main/java/redelm/io/BaseRecordReader.java [hunk 6]
```diff
@@ -0,0 +1,151 @@
...
+
+public abstract class BaseRecordReader<T> extends RecordReader<T> {
+  private static final Log LOG = Log.getLog(BaseRecordReader.class);
+
+  public RecordConsumer recordConsumer;
+  public RecordMaterializer<T> recordMaterializer;
+  public ColumnsStore columnStore;
+  @Override
+  public T read() {
+    readOneRecord();
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/io/BaseRecordReader.java:31]: Same comment as elsewhere. You can collapse this pattern by making readOneRecord() return the current record, since this is how you use it everywhere

File: redelm-column/src/main/java/redelm/io/BaseRecordReader.java [hunk 7]
```diff
@@ -0,0 +1,151 @@
...
+    return recordMaterializer.getCurrentRecord();
+  }
+
+  /* (non-Javadoc)
+   * @see redelm.io.RecordReader#read(T[], int)
+   */
+  @Override
+  public void read(T[] records, int count) {
+    if (count > records.length) {
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/io/BaseRecordReader.java]: If you're going to have the check ,you might as well include some more info (ie the count, the records size, and so on)

File: redelm-column/src/main/java/redelm/io/BaseRecordReader.java [hunk 8]
```diff
@@ -0,0 +1,151 @@
...
+  protected abstract void readOneRecord();
+
+  State[] caseLookup;
+
+  private String endField;
+
+  private int endIndex;
+
+  protected void currentLevel(int currentLevel) {
+    if (DEBUG) LOG.debug("currentLevel: "+currentLevel);
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/io/BaseRecordReader.java:44]: I'm going to register my final complain about the if (DEBUG) style then will shut up

File: redelm-column/src/main/java/redelm/io/BaseRecordReader.java [hunk 9]
```diff
@@ -0,0 +1,151 @@
...
+  private void endField(String field, int index) {
+    if (DEBUG) LOG.debug("endField("+field+","+index+")");
+    if (endField != null) {
+      recordConsumer.endField(endField, endIndex);
+    }
+    endField = field;
+    endIndex = index;
+  }
+
+  final protected void addPrimitiveSTRING(String field, int index, String value) {
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/io/BaseRecordReader.java]: A healthy distrust of method overloading?

File: redelm-column/src/main/java/redelm/io/ColumnIO.java [hunk 10]
```diff
@@ -43,10 +43,18 @@
     return fieldPath;
   }
 
+  public String getFieldPath(int level) {
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/io/ColumnIO.java:46]: Sometimes you have an accessor method like this (which I support), but then sometimes you have stuff like:
array[i].method(dog)[method2()].method3() which can be a bit overwhelming. It might be worth breaking out some of those into methods like this so we can document the purpose? Or maybe just documenting them in general, as it can be difficult to fully understand them.

File: redelm-column/src/main/java/redelm/io/RecordReaderCompiler.java [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+      return super.defineClass(type.getName(), classBytes, 0, classBytes.length);
+    }
+  }
+
+  private DynamicClassLoader cl = new DynamicClassLoader();
+  private int id = 0;
+
+  private <S extends StatementBuilder<S>> S generateCase(boolean addPrimitive, State state, Case currentCase, S builder, int stateCount) {
+    if (currentCase.isGoingUp()) {
+      //  for (; currentLevel <= depth; ++currentLevel) {
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/io/RecordReaderCompiler.java:62]: is there a reason this commented out code is here, and elsewhere, and can't be deleted?

File: redelm-column/src/main/java/redelm/io/RecordReaderImplementation.java [hunk 12]
```diff
@@ -0,0 +1,498 @@
...
+    private final boolean goingUp;
+    private final boolean goingDown;
+    private final int nextState;
+
+    public Case(int startLevel, int depth, int nextLevel, int nextState) {
+      this.startLevel = startLevel;
+      this.depth = depth;
+      this.nextLevel = nextLevel;
+      this.nextState = nextState;
+      goingUp = startLevel <= depth;
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/io/RecordReaderImplementation.java:62]: The following comments would be useful: 1. What does going up vs going down mean 2. In the above heavy brennus stuff, some general comments guiding what each piece of code is meant to do.

File: redelm-column/src/main/java/redelm/io/RecordReaderImplementation.java [hunk 13]
```diff
@@ -0,0 +1,498 @@
...
+      goingDown = depth + 1 > nextLevel;
+    }
+
+    public void setID(int id) {
+      this.id = id;
+    }
+
+    @Override
+//    public int hashCode() {
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/io/RecordReaderImplementation.java:76]: Same comments re: commented code apply to this file

File: redelm-column/src/main/java/redelm/io/RecordReaderImplementation.java [hunk 14]
```diff
@@ -0,0 +1,498 @@
...
+//        hashCode += 1 * (1 + startLevel) + 2 * (1 + depth);
+//      }
+//      if (goingDown) {
+//        hashCode += 3 * (1 + depth) + 5 * (1 + nextLevel);
+//      }
+//      return hashCode;
+//    }
+
+    public int hashCode() {
+      return 1 * startLevel + 2 * depth + 3 * nextLevel + 5 * nextState;
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/io/RecordReaderImplementation.java]: I think the following has better mathematical properties, but obviously it's all heuristics.

31 \* (31 \* ( 31 \* (31 \* (17 + startLevel) + depth) + nextLevel) + nextState);

This is just going off of some old readings of Josh Bloch et al

File: redelm-column/src/main/java/redelm/io/RecordReaderImplementation.java [hunk 15]
```diff
@@ -0,0 +1,498 @@
...
+      if (d >= currentState.maxDefinitionLevel) {
+        // not null
+        addPrimitive(currentState, columnReader);
+      }
+      columnReader.consume();
+
+      int nextR = currentState.maxRepetitionLevel == 0 ? 0 : columnReader.getCurrentRepetitionLevel();
+      // level to go to close current groups
+      int next = currentState.nextLevel[nextR];
+      for (; currentLevel > next; currentLevel--) {
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/io/RecordReaderImplementation.java:385]: I prefer do while to the for(;) type (ie do {} while (currentLevel-- > next) though am fine either way.

File: redelm-column/src/main/java/redelm/io/RecordReaderImplementation.java [hunk 16]
```diff
@@ -0,0 +1,498 @@
...
+    return nextState == null ? states.length : nextState.id;
+  }
+
+  int getNextLevel(int current, int nextRepetitionLevel) {
+    return states[current].nextLevel[nextRepetitionLevel];
+  }
+
+  private int getCommonParentLevel(String[] previous, String[] next) {
+    int i = 0;
+    while (i < previous.length && i < next.length && previous[i].equals(next[i])) {
```
Comment (jcoveney) [inline redelm-column/src/main/java/redelm/io/RecordReaderImplementation.java]: while (i < Math.min(previous.length, next.length) && ...

File: redelm-column/src/test/java/redelm/io/PerfTest.java [hunk 17]
```diff
@@ -40,39 +39,49 @@
   public static void main(String[] args) {
     MemColumnsStore columns = new MemColumnsStore(50*1024*1024, schema);
     write(columns);
-    columns.flip();
+//    columns.flip();
```
Comment (jcoveney) [inline redelm-column/src/test/java/redelm/io/PerfTest.java]: Flip was always somewhat confusing. Should it be in, or out?

File: redelm-column/src/main/java/redelm/io/BaseRecordReader.java (hunk 9 above)
Comment (julienledem) [inline redelm-column/src/main/java/redelm/io/BaseRecordReader.java]: if it is final it will use INVOKESPECIAL instead of INVOKEVIRTUAL I was trying to see if it would be faster that way.
INVOKEVIRTUAL does an extra lookup to support overloading.

File: redelm-column/src/main/java/redelm/io/BaseRecordReader.java (hunk 8 above)
Comment (julienledem) [inline redelm-column/src/main/java/redelm/io/BaseRecordReader.java:44]: if there's no "if (DEBUG)" it will concatenate the string even if LOG.debug does nothing.
"currentLevel: "+currentLevel is evaluated which involves copying bytes around.
"if (DEBUG)" guarantees the line disappears altogether. Other optimizations at runtime are unclear.

File: redelm-column/src/main/java/redelm/io/BaseRecordReader.java (hunk 6 above)
Comment (julienledem) [inline redelm-column/src/main/java/redelm/io/BaseRecordReader.java:31]: readOneRecord is generated, I was simplifying the generation by avoiding the generic bit.

File: redelm-column/src/main/java/redelm/io/ColumnIO.java (hunk 10 above)
Comment (julienledem) [inline redelm-column/src/main/java/redelm/io/ColumnIO.java:46]: If you could point out specific lines that would be helpful.

File: redelm-column/src/main/java/redelm/io/RecordReaderCompiler.java (hunk 11 above)
Comment (julienledem) [inline redelm-column/src/main/java/redelm/io/RecordReaderCompiler.java:62]: This is not very clear but this is the bit of code from the original code that is being generated here.

File: redelm-column/src/main/java/redelm/io/RecordReaderImplementation.java (hunk 12 above)
Comment (julienledem) [inline redelm-column/src/main/java/redelm/io/RecordReaderImplementation.java:62]: Yes, I need to improve this Case class which is there to deduplicate what is done with the various combinations of those variables.

File: redelm-column/src/main/java/redelm/io/RecordReaderImplementation.java (hunk 13 above)
Comment (julienledem) [inline redelm-column/src/main/java/redelm/io/RecordReaderImplementation.java:76]: This was buggy but I need to move back to something closer to this as the current hashCode()/equals() are not precise enough

File: redelm-column/src/test/java/redelm/io/PerfTest.java (hunk 17 above)
Comment (julienledem) [inline redelm-column/src/test/java/redelm/io/PerfTest.java]: The flip thing needs to be reworked but that will be for another PR

File: redelm-column/src/main/java/redelm/io/BaseRecordReader.java (hunk 7 above)
Comment (julienledem) [inline redelm-column/src/main/java/redelm/io/BaseRecordReader.java]: done
